sycl/source/detail/stream_impl.cpp

```diff
@@ -59,21 +61,33 @@ size_t stream_impl::get_size() const { return BufferSize_; }
 
 size_t stream_impl::get_max_statement_size() const { return MaxStatementSize_; }
 
-void stream_impl::flush() {
-  // Access the stream buffer on the host. This access guarantees that kernel is
-  // executed and buffer contains streamed data.
-  {
-    auto HostAcc = detail::Scheduler::getInstance()
-                       .StreamBuffersPool.find(this)
-                       ->second.Buf.get_access<cl::sycl::access::mode::read>(
-                           range<1>(BufferSize_), id<1>(OffsetSize));
+std::once_flag flag;
 
-    printf("%s", HostAcc.get_pointer());
-    fflush(stdout);
-  }
+void stream_impl::flush() {
+  // We don't want stream flushing to be blocking operation that is why submit a
+  // host task to print stream buffer.
+  queue Q = detail::createSyclObjFromImpl<queue>(
+      cl::sycl::detail::Scheduler::getInstance().getDefaultHostQueue());
+  Q.submit([&](handler &cgh) {
+    auto HostAcc =
+        detail::Scheduler::getInstance()
+            .StreamBuffersPool.find(this)
+            ->second.Buf
+            .get_access<access::mode::read, access::target::host_buffer>(
+                cgh, range<1>(BufferSize_), id<1>(OffsetSize));
+    cgh.codeplay_host_task([=]() {
+      printf("%s", HostAcc.get_pointer());
+      fflush(stdout);
+    });
+  });
 
-  // Flushed the stream, can deallocate the buffers now.
-  detail::Scheduler::getInstance().deallocateStreamBuffers(this);
+  // Register a library cleanup function to deallocate stream buffers, we need
+  // to do it only once. Deallocation of the buffer will also guarantee that
+  // copy back and printing is completed.
+  std::call_once(flag, []() {
+    std::atexit(
```

Comment:
Should we check for non-zero return code and have a workaround upon non-zero value?
C++ std guarantees only some amount of registered functions.